text/0255-sensitive-data-redaction.md

```diff
@@ -0,0 +1,397 @@
+
+The `DROP` keyword means that the value is replaced with `REDACTED` (or set to null, …).
+
+It is the responsibility of the OpenTelemetry SDK to implement those sensitivity details provided by the semantic conventions.
+
+### Annotate attributes with sensitivity information
+
+As an additional building block the OpenTelemetry API should provide capabilities
+to library (and application) authors to apply in-place redaction
+
+One option is to add additional attributes to methods like `setAttribute`:
+
+```
+span.setAttribute("url.query", url.toString(), <SENSITIVITY_DETAILS>);
+```
+
+Or, if the method can not be extended, an additional method can be called:
+
+```
+span.setAttribute("url.query", OpenTelemetry.redact(url.toString(), <SENSITIVITY_DETAILS>);
+```
+
+Note that those API calls are no-op and will be implemented by the SDK (as we do it with other API methods as well), this way (almost?) no additional overhead will be created by introducing those annotations.
+
+The `<SENSITIVY_DETAILS>` should look like the following:
+
+```
+{
+  `<REDACTION_PROFILE>` => `<REDACTION_CONFIGURATION>`,
+  ...
+}
+```
+
+There may be multiple redaction profiles (like `DEFAULT`, `STRICT`, etc.), and
+for each one a configuration (as outlined above) may be applied.
+
+any sensitive data.
+
+## Internal details
+
+### Redactor
+
+To accomplish redaction the SDK needs a component (similar to a sampler) that inspects attributes when they are set and applies the required redactions:
+
+- `Redactor::setup(profile)` will setup the redactor with the given profile,
+  maybe a constructor? depends on the language
+- `Redactor::redact(value, <sensitivityDetails>)` will return the redacted value
+  using the provided profile
+
+The redactor will also have all the methods to apply predefined redactions
+(`REDACT_ALL_URL_VALUES`, `REDACT_INSECURE_URL_PARAM_VALUES`, `DROP`, etc.).
+If a method is not implemented (either by the SDK or by the end-user choosing
+one that does not exist), it will default to apply `DROP` to avoid leakage of
+
+## Additional context
+
+Treating sensitive data properly is a very complex multi-dimensional topic.
+Below you will find some additional context on that subject.
+
+### Types of sensitive data
+
+There are different kinds of “sensitivity” that may apply to data. The ones most
+relevant in this proposal are “security” and “privacy”. They may overlap but we
+distinguish them as follows:
+
+- security-relevant sensitive data: any information that when exposed
+  [weakens the overall security of the device/system](https://en.wikipedia.org/wiki/Vulnerability_(computing)).
+- privacy-relevant sensitive data: any information that when exposed
+  [can adversely affect the privacy or welfare of an individual](https://en.wikipedia.org/wiki/Information_sensitivity).
+
```

Comment:
The identification of sensitive data could also be applied at a semantic level. For example, as a company, I might want to apply a redaction rule to ALL URLs, another rule to ALL IP addresses, and so on. One possible approach would be to introduce a new type of group in the semconv that would allow the definition of semantic types, which could be attached to existing attributes in addition to the "physical" data type that we already have. The redaction rules would then apply to the semantic types.